src/rbac.rs

```diff
@@ -0,0 +1,385 @@
+        for rule in self.groups.iter() {
+            // If ANY rule matches, it is a match...
+            let mut rule_match = true;
+            for group in rule.iter() {
+                // We need ALL groups to match...
+                let mut group_match = true;
+                for mg in group.iter() {
+                    if mg.is_empty() {
+                        trace!(matches = false, "empty rule");
+                        group_match = false;
+                        break;
+                    }
+                    // We need ALL of these to match. Within each type, ANY must match
+                    let mut m = true;
+                    m &= Self::matches_internal(
+                        "destination_ip",
+                        &mg.destination_ips,
+                        &mg.not_destination_ips,
+                        |i| i.contains(&conn.dst.ip()),
+                    );
+                    m &= Self::matches_internal(
+                        "source_ips",
+                        &mg.source_ips,
+                        &mg.not_source_ips,
+                        |i| i.contains(&conn.src_ip),
+                    );
+                    m &= Self::matches_internal(
+                        "destination_ports",
+                        &mg.destination_ports,
+                        &mg.not_destination_ports,
+                        |p| *p == conn.dst.port(),
+                    );
+                    m &= Self::matches_internal(
+                        "principals",
+                        &mg.principals,
+                        &mg.not_principals,
+                        |p| p.matches(&id),
+                    );
+                    m &= Self::matches_internal(
+                        "namespaces",
+                        &mg.namespaces,
+                        &mg.not_namespaces,
+                        |p| p.matches(&ns),
+                    );
+
+                    group_match &= m;
+                }
+
+                rule_match &= group_match;
+            }
+            if rule_match {
+                return true;
+            }
+        }
+        false
+    }
+
+    #[instrument(name= "match", level = "trace", skip_all, fields(%desc))]
+    fn matches_internal<T: fmt::Debug>(
+        desc: &'static str,
+        positive: &Vec<T>,
+        negative: &Vec<T>,
+        mut predicate: impl FnMut(&T) -> bool,
+    ) -> bool {
+        let pm = if positive.is_empty() {
+            trace!(matches = true, "type" = "positive", "no match declared");
+            true
+        } else {
+            let matches = positive.iter().any(&mut predicate);
+            trace!(%matches, "type"="positive", "{positive:?}");
+            matches
+        };
+        let nm = if negative.is_empty() {
+            trace!(matches = true, "type" = "negative", "no match declared");
+            true
+        } else {
+            let matches = negative.iter().any(&mut predicate);
```

Comment:
Negate predicate this or change `any` to `none`

Reply:
or I guess just make it check `pm && !nm` is cleaner

Reply:
good catch - will add a bunch of tests before we merge this